### PR TITLE
[vdk-core] Implementation of new ingestion interfaces

### PR DESCRIPTION
**Pull request by doks5**
As part of https://github.com/vmware/versatile-data-kit/pull/682
and https://github.com/vmware/versatile-data-kit/pull/672 the ingestion
interfaces were extended to support ingestion plugin chaining for pre- and
post- processing operations.

This change introduces the implementation of the extended ingestion functionality.
The IngesterRouter and IngesterBase classes are extended to support the use of pre-
and post- ingestion processing plugins.

Testing Done: unit tests; functional tests to follow in a separate PR.

Signed-off-by: Andon Andonov <andonova@vmware.com>

### CODE REVIEW

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py [hunk 1]
```diff
@@ -41,6 +42,16 @@ def __init__(self, cfg: Configuration, core_state: StateStore):
...
 
+        self._pre_processor_sequence: Optional[List] = self.__get_plugin_sequence(
+            "INGEST_PAYLOAD_PREPROCESS_SEQUENCE"
+        )
+        self._post_processor_sequence: Optional[List] = self.__get_plugin_sequence(
+            "INGEST_PAYLOAD_POSTPROCESS_SEQUENCE"
+        )
+
+        self._initialized_pre_processors: List[Optional[IIngesterPlugin]] = []
+        self._initialized_post_processors: List[Optional[IIngesterPlugin]] = []
```
Comment (ivakoleva) [inline projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py:53]: I don't understand why adding four fields;
`self._ingester_builders` are populated with all ingested plugins already, what is needed additionally,
is literally "give me pre-processors out of those, then filter by config-provided pre-sequence",
also "give me post-processors out of those, then filter by config-provided post-sequence".

I would consider this a short lambda potentially. And even name of pre_ingest_process/post_ingest_process along with INGEST_PAYLOAD_PREPROCESS_SEQUENCE/INGEST_PAYLOAD_POSTPROCESS_SEQUENCE could be parameterized, so one lambda is needed for both similar cases.

Then, when evaluating the processors 
-> common to ingestion itself, where method configured is missing raise misconfiguration exception
-> if number of filtered pre-processors differs than config.split then raise exception
-> if number of filtered post-...same

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py [hunk 2]
```diff
@@ -226,10 +250,50 @@ def __initialize_ingester(self, method) -> IngesterBase:
                 self._state.get(CommonStoreKeys.OP_ID),
                 ingester_plugin,
                 IngesterConfiguration(config=self._cfg),
+                pre_processors=self._initialized_pre_processors,
+                post_processors=self._initialized_post_processors,
             )
 
         return self._cached_ingesters[method]
 
+    def __get_plugin_sequence(self, key: str) -> List:
```
Comment (ivakoleva) [inline projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py:259]: if suggestion applied, then this 259-296 is redundant;

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py [hunk 3]
```diff
@@ -59,6 +70,12 @@ def send_object_for_ingestion(
         target: Optional[str] = None,
         collection_id: Optional[str] = None,
     ):
+        # Initialize the pre- and post- processors.
```
Comment (ivakoleva) [inline projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py:73]: filter by lambda, then check if processors could match the configured, else raise an exception;
take a loot at `__initialize_ingester` to reuse some parts of it?maybe simplify it to apply to all cases? could lets sync on this one to elaborate on options if needed

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py [hunk 4]
```diff
@@ -45,6 +45,8 @@ def __init__(
         op_id: str,
         ingester: IIngesterPlugin,
         ingest_config: IngesterConfiguration,
+        pre_processors: List[Optional[IIngesterPlugin]] = [],
```
Comment (ivakoleva) [inline projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py:48]: default initialization seems redundant, and also IDE displays messages

File: projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_router.py [hunk 5]
```diff
@@ -76,48 +77,31 @@ def test_router_send_tabular_data_for_ingestion_no_default_method(
...
 @patch(f"{IngesterRouter.__module__}.{IngesterBase.__name__}", spec=IngesterBase)
-def test_router_use_chained_ingest_plugins_send_object(
-    mock_ingester_base: MagicMock,
-):
-    router = create_ingester_router({"ingest_method_default": "test-ingest"})
-
-    router.add_ingester_factory_method(
-        "pre-ingest-test", lambda: MagicMock(spec=IIngesterPlugin)
-    )
-    router.add_ingester_factory_method(
```
Comment (ivakoleva) [inline projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_router.py:87]: I like you had registered two infesters, then one is configured for usage via  `ingest_method_default`, 
and the verification is no chained infesters while multiple ones registered in the first place

File: projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_router.py [hunk 6]
```diff
@@ -76,48 +77,31 @@ def test_router_send_tabular_data_for_ingestion_no_default_method(
...
+    router = create_ingester_router(config)
 
-    mock_ingester_base.return_value.send_tabular_data_for_ingestion.assert_called_with(
-        ["b"], ["a"], None, "test-ingest", None, None
-    )
-    # TODO: Uncomment when implementation is ready
-    # mock_ingester_base.return_value.send_tabular_data_for_ingestion.assert_called_with(
-    #     ["b"], ["a"], None, "pre-ingest-test", None, None
-    # )
+    with pytest.raises(VdkConfigurationError):
```
Comment (ivakoleva) [inline projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_router.py:103]: maybe add check the processors are now initialised once configured?

File: projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_router.py [hunk 7]
```diff
@@ -76,48 +77,31 @@ def test_router_send_tabular_data_for_ingestion_no_default_method(
...
+        router.send_object_for_ingestion({"a": "b"})
+
+    with pytest.raises(VdkConfigurationError):
```
Comment (ivakoleva) [inline projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_router.py:106]: maybe check also the config error nature, like message expected

File: projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_base.py [hunk 8]
```diff
@@ -27,12 +26,21 @@ def create_ingester_base() -> IngesterBase:
     }
     test_config = Configuration(None, config_key_value_pairs, {})
 
-    return IngesterBase(
-        data_job_name="test_job",
-        op_id="42a420",
-        ingester=MagicMock(spec=IIngesterPlugin),
-        ingest_config=IngesterConfiguration(test_config),
-    )
```
Comment (ivakoleva) [inline projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_base.py:35]: ```**(kwargs or {})```
?
then remove if-else redundant

File: projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_base.py [hunk 9]
```diff
@@ -41,7 +49,7 @@ def test_send_object_for_ingestion(mocked_send):
     destination_table = "a_destination_table"
     method = "test_method"
     target = "some_target"
-    ingester_base = create_ingester_base()
+    ingester_base = create_ingester_base(None)
```
Comment (ivakoleva) [inline projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_base.py:52]: maybe def ```create_ingester_base(kwargs={})``` to simplify even more?

**Review by ivakoleva — CHANGES_REQUESTED** [review]
missing test_ignester_base.py scenarios about:
1. pre ingestion
2. payload modification propagation from pre-processing to ingestion
3. metadata propagation verification when chaining (from pre, to ingest, to post -processing)
4. multiple pre-ingest/post-ingest triggered in a particular sequence

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py [hunk 10]
```diff
@@ -521,18 +547,17 @@ def _payload_poster_thread(self):
...
 
-            # Complete Post-Ingestion operations
-            try:
-                self._ingester.post_ingest_process(
-                    payload=payload_dict,
+            # If there are any post-processors set, complete the post-process
+            # operations
+            if self._post_processors:
+                self._execute_post_process_operations(
+                    payload=payload_obj,
```
Comment (ivakoleva) [inline projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py:554]: I was looking into post processor implementation, and did notice the post_ingest_process() payload parameter is optional? why is that?

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py [hunk 11]
```diff
@@ -575,6 +600,74 @@ def close_now(self):
...
+            try:
+                metadata = plugin.post_ingest_process(
+                    payload=payload,
+                    destination_table=destination_table,
+                    target=target,
+                    collection_id=collection_id,
+                    metadata=metadata,
+                    exception=exception,
+                )
+            except Exception as e:
```
[the remark this answers is not in the document]
Comment (doks5) [reply projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py:658]: Good point. Added to exception message, thanks

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py (hunk 3 above)
Comment (doks5) [reply projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py:66]: Refactored to be cleaner.

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py (hunk 2 above)
Comment (doks5) [reply projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py:274]: Removed some unnecessary bridge methods. Thanks

File: projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_base.py (hunk 8 above)
Comment (doks5) [reply projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_base.py:35]: Doesn't work, IngesterBase expects either pre-processors and post-processors, or nothing. An empty dictionary is still considered something by python, so using `**(kwargs or {})` throws an exception.

File: projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_router.py (hunk 5 above)
Comment (doks5) [reply projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_router.py:87]: Expanded one of the tests to incorporate this suggestion. Thank you

File: projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_router.py (hunk 6 above)
Comment (doks5) [reply projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_router.py:107]: The processors would not be initialized in this case, as there are no registered plugins.

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py [hunk 12]
```diff
@@ -575,6 +600,74 @@ def close_now(self):
...
+        for plugin in self._pre_processors:
+            try:
+                payload, metadata = plugin.pre_ingest_process(
+                    payload=payload,
+                    destination_table=destination_table,
+                    target=target,
+                    collection_id=collection_id,
+                    metadata=metadata,
+                )
+            except:
```
[the remark this answers is not in the document]
Comment (doks5) [reply projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py]: Fixed, thanks

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py [hunk 13]
```diff
@@ -472,19 +483,34 @@ def _payload_poster_thread(self):
...
+                if self._pre_processors:
+                    payload_obj, ingestion_metadata = self._pre_process_payload(
+                        payload=payload_obj,
+                        destination_table=destination_table,
+                        target=target,
+                        collection_id=collection_id,
+                        metadata=ingestion_metadata,
+                    )
 
                 try:
```
[the remark this answers is not in the document]
Comment (doks5) [reply projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py]: Refactored, thanks for the suggestions

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py (hunk 1 above)
Comment (doks5) [reply projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py]: Removed unnecessary fields, and did some refactoring.

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py (hunk 4 above)
Comment (doks5) [reply projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py]: Fixed, thanks for pointing it out

File: projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_router.py (hunk 7 above)
Comment (doks5) [reply projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_router.py]: Done, thanks

File: projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_base.py (hunk 9 above)
Comment (doks5) [reply projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_base.py]: Did something similar, as the IDE did not like `kwargs={}`.

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py (hunk 10 above)
Comment (doks5) [reply projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py]: Refactored.

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py [hunk 14]
```diff
@@ -226,10 +265,44 @@ def __initialize_ingester(self, method) -> IngesterBase:
                 self._state.get(CommonStoreKeys.OP_ID),
                 ingester_plugin,
                 IngesterConfiguration(config=self._cfg),
+                pre_processors=self._initialized_pre_processors,
+                post_processors=self._initialized_post_processors,
             )
 
         return self._cached_ingesters[method]
 
+    def __get_plugin_sequence(self, key: str) -> List:
```
Comment (ivakoleva) [inline projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py:274]: seems like a config utility method, for parsing configuration sequences by delimiter;

File: projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_base.py [hunk 15]
```diff
@@ -27,12 +26,21 @@ def create_ingester_base() -> IngesterBase:
...
+            ingest_config=IngesterConfiguration(test_config),
+            **kwargs
+        )
+    else:
+        return IngesterBase(
+            data_job_name="test_job",
+            op_id="42a420",
+            ingester=MagicMock(spec=IIngesterPlugin),
+            ingest_config=IngesterConfiguration(test_config),
+        )
```
Comment (ivakoleva) [inline projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_base.py:43]: interesting; is it possible to optimise like
if in favour of None:
```def create_ingester_base(kwargs={}) -> IngesterBase:```

then
```suggestion
    return IngesterBase(
        data_job_name="test_job",
        op_id="42a420",
        ingester=MagicMock(spec=IIngesterPlugin),
        ingest_config=IngesterConfiguration(test_config),
        **kwargs
    )
```

File: projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_router.py (hunk 6 above)
Comment (ivakoleva) [reply projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_router.py:107]: yup, seems missing coverage for scenarios when processors are initialized, pre- and post-

**Review by ivakoleva — CHANGES_REQUESTED** [review]
maybe add router tests to verify processors initialisation, then merge?

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py [hunk 16]
```diff
@@ -111,9 +132,27 @@ def send_tabular_data_for_ingestion(
...
+                self.__initialize_processor(i) for i in pre_processor_sequence
+            ]
+
+        if not self._initialized_post_processors:
+            post_processor_sequence: Optional[List] = self.__get_plugin_sequence(
+                "INGEST_PAYLOAD_POSTPROCESS_SEQUENCE"
+            )
+            self._initialized_post_processors = [
+                self.__initialize_processor(i) for i in post_processor_sequence
+            ]
```
[the remark this answers is not in the document]
Comment (doks5) [reply projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py]: Refactored, thanks

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py (hunk 14 above)
Comment (doks5) [reply projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py]: Indeed. Moved to utils.

File: projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_base.py (hunk 15 above)
Comment (doks5) [reply projects/vdk-core/tests/vdk/internal/builtin_plugins/ingestion/test_ingester_base.py]: Setting `kwargs={}` in the method signature was causing issues, as dictionaries are mutable objects, but with a slight workaround it was indeed possible to get rid of the duplicate code. Thank you for the suggestion

**Comment by doks5** [comment]
> maybe add router tests to verify processors initialisation, then merge?

@ivakoleva 
Test added. Thanks for the review

**Comment by ivakoleva** [comment]
> > maybe add router tests to verify processors initialisation, then merge?
> 
> @ivakoleva Test added. Thanks for the review

couldn't find pre-processors tests, sorry if I missed them by any chance,
what I found on  processors coverage,
`test_ingester_router.py` covers:
- test_router_chained_ingest_plugins <- pre&post processors count initialised corresponds to configured count

`test_ingester_base.py` covers:
- test_ingest_payload_and_post_ingestion_operation <- post processor evaluation with parameters expected, in case no exception
- test_post_ingestion_operation_with_exceptions <- post processor evaluation with parameters expected, in case exception

**Comment by doks5** [comment]
> > > maybe add router tests to verify processors initialisation, then merge?
> > 
> > 
> > @ivakoleva Test added. Thanks for the review
> 
> couldn't find pre-processors tests, sorry if I missed them by any chance, what I found on processors coverage, `test_ingester_router.py` covers:
> 
>     * test_router_chained_ingest_plugins <- pre&post processors count initialised corresponds to configured count
> 
> 
> `test_ingester_base.py` covers:
> 
>     * test_ingest_payload_and_post_ingestion_operation <- post processor evaluation with parameters expected, in case no exception
> 
>     * test_post_ingestion_operation_with_exceptions <- post processor evaluation with parameters expected, in case exception

@ivakoleva 
Added tests in test_ingester_base to verify the operations of pre-processors. Thanks

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py [hunk 17]
```diff
@@ -59,6 +64,17 @@ def send_object_for_ingestion(
         target: Optional[str] = None,
         collection_id: Optional[str] = None,
     ):
+        # Initialize the pre- and post- processors.
+        if not self._initialized_pre_processors:
```
Comment (ivakoleva) [inline projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py:68]: why initialising processors on each send_*_for_ingestion?
why not when initialising the ingesterbase in `__initialize_ingester()` where they are used?

**Review by ivakoleva — CHANGES_REQUESTED** [review]

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py (hunk 17 above)
Comment (doks5) [reply projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py:68]: @ivakoleva 
They are not initialized on each send_*_for_ingestion, they are initialized only on the first invocation of a send_*_for_ingestion method, and then reused.
Moving the initialization to where the IngesterBase is initialized does not seem to provide any significant benefit.

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py [hunk 18]
```diff
@@ -59,6 +64,17 @@ def send_object_for_ingestion(
...
+            self._initialized_pre_processors = self.__get_initialized_processors(
+                "INGEST_PAYLOAD_PREPROCESS_SEQUENCE"
+            )
+
+        if not self._initialized_post_processors:
+            self._initialized_post_processors = self.__get_initialized_processors(
```
Comment (ivakoleva) [inline projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py:74]: I see one more initialisation here, on each `send_tabular_data_for_ingestion()`,

yet both send_*_for_ingestion() methods use cached infesters if available, therefore initialising the ingesterbase once - yet processors are reinitialised each time it seems to me?

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py [hunk 19]
```diff
@@ -111,9 +127,21 @@ def send_tabular_data_for_ingestion(
...
-        # if set. `method` and `target` provided when the method is called take
-        # precedence over the ones set with environment variables.
+        # Initialize the pre- and post- processors.
+        if not self._initialized_pre_processors:
+            self._initialized_pre_processors = self.__get_initialized_processors(
+                "INGEST_PAYLOAD_PREPROCESS_SEQUENCE"
+            )
+
+        if not self._initialized_post_processors:
+            self._initialized_post_processors = self.__get_initialized_processors(
```
[the remark this answers is not in the document]
Comment (doks5) [reply projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py]: Fixed, thanks

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py (hunk 18 above)
Comment (doks5) [reply projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py]: Fixed, thanks

**Review by ivakoleva — APPROVED** [review]

File: projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py [hunk 20]
```diff
@@ -221,15 +224,57 @@ def __initialize_ingester(self, method) -> IngesterBase:
                 f"If upgraded recently consider reverting to previous version. Or use another method type.",
             )
         else:
+            # Initialize the pre- and post- processors.
+            initialized_pre_processors = self.__get_initialized_processors(
```
Comment (ivakoleva) [inline projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py:228]: why not initialize processors once? the config sequences wouldn't change in runtime,
yet multiple ingesters could be initialized

Comment (ivakoleva) [reply projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_router.py:228]: got the idea; I'm ok with that impl

**Pull request merged**